### DISCUSSION TITLE
Debug C++ in Pulsar (LLDB preferred)

### DISCUSSION THREAD

**Discussion by b3n-h4il** [pulsar-edit/.github#284] [Community Package Help] [answered]
Is there someone also using Pulsar for C++ development (mostly learning in my case)? I'd like some pointer on how to debug code. Searching the packages, I've found some debug extensions, but they're not maintained and use gdb. I'm willing to learn how to develop an extension to do that, but at this stage it'd be a long way for me.

**Comment by savetheclocktower** [comment]
The bad news is that I'm not aware of any extension that acts as a frontend to a C/C++ debugger — though we'd certainly help with mentoring if you wanted to modernize one of those unmaintained extensions.

Despite the fact that it doesn't do debugging, the best option is probably [pulsar-ide-clangd](https://web.pulsar-edit.dev/packages/pulsar-ide-clangd) — which uses the [clangd](https://clangd.llvm.org/) language server:

* Out of the box you'll get autocompletion and symbol navigation (via <kbd>Cmd-R</kbd>/<kbd>Ctrl-R</kbd>) and the ability to jump to the declaration of the symbol under the cursor (the **Symbols View: Go To Declaration** command, whose hotkey varies by platform).
* Combine it with [linter](https://web.pulsar-edit.dev/packages/linter) and [linter-ui-default](https://web.pulsar-edit.dev/packages/linter-ui-default) and you'll get annotation of errors/warnings in your code.
* Other packages listed in `pulsar-ide-clangd`’s README deliver userful features; you can pick and choose which ones you want.

**Reply by b3n-h4il** [reply]
Thank you for your quick answer. Yeah, I'm using those extensions, and I really like them, you're their maintainer right? Thanks again. The mentoring to create an extension for a front-end for the debugger sounds interesting, but currently I don't have a lot of knowledge about web tech, and understand very litle of how Pulsar works under the hood. Other than the Launch manual, do you have a suggestion for a beginner developing an extension? I've been search for some tutorial to create a "Hello world" kinda extension, but no success yet.

**Reply by savetheclocktower** [reply, marked as answer]
> Other than the Launch manual, do you have a suggestion for a beginner developing an extension? I've been search for some tutorial to create a "Hello world" kinda extension, but no success yet.

The [“Word Count” package tutorial](https://docs.pulsar-edit.dev/developing-for-pulsar/package-word-count/) in the docs is the closest thing we have. (That URL links to our new docs, which are not 100% done but are hopefully easier to read and navigate through.) There are a couple tutorials that follow for other sorts of packages.

If you get through those and have an idea for _any other_ sort of package _other than_ a debugger, I'd do that one first, because a debugger is jumping straight from Algebra 101 to upper-division Calculus. (Though, of course, it's not quite as intimidating if you're trying to rehabilitate an old extension that used to work.)

**Reply by savetheclocktower** [reply]
> Yeah, I'm using those extensions, and I really like them, you're their maintainer right? Thanks again.

You're welcome! I do maintain most of the `pulsar-ide-` packages, but most of those are just modernizations of work done by other people. I didn't change much about `pulsar-ide-clangd` except update one of its dependencies to improve integration with things that we've added to Pulsar since the fork.

**Comment by b3n-h4il** [comment]
Something else: Is this the best place to talk about extension development and Pulsar stuff? The Lemmy account seems not as popular as the GitHub discussions. I'm not a huge fan of talking one-to-one personally, I like foruns better.

**Reply by savetheclocktower** [reply]
For something like this, probably. The Discord sees the most activity out of anything, but if you prefer forums to chat, then this is it.
